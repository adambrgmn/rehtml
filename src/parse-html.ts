import { parseFragment } from 'parse5';

interface TextNode {
  type: 'text';
  data: string;
}

interface TagNode {
  type: 'tag';
  name: keyof JSX.IntrinsicElements;
  props: Record<string, string | boolean>;
  children: (TextNode | TagNode)[];
}

type DOMNode = TextNode | TagNode;
type DOMTree = DOMNode[];

const parseHtml = (html: string) => {
  const parsed = parseFragment(html);

  const reduceChildren = (nodes: any[]): DOMTree =>
    nodes.reduce((tree: DOMTree, node: any) => {
      switch (node.nodeName) {
        case '#text':
          const textNode: TextNode = {
            type: 'text',
            data: node.value as string,
          };

          return [...tree, textNode];

        default:
          const tagNode: TagNode = {
            type: 'tag',
            name: node.tagName as keyof JSX.IntrinsicElements,
            props: node.attrs.reduce(
              (
                acc: Record<string, string>,
                { name, value }: { name: string; value: string },
              ) => ({ ...acc, [name]: value }),
              {},
            ),
            children: reduceChildren(node.childNodes),
          };
          return [...tree, tagNode];
      }
    }, []);

  // @ts-ignore
  const nodes = reduceChildren(parsed.childNodes);
  return nodes;
};

export { parseHtml };
