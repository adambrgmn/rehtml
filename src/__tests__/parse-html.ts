import { parseHtml } from '../parse-html';

it('should parse an html string into a DOMTree', () => {
  const html = '<p class="foo">Hello world</p>';
  const result = parseHtml(html);

  expect(result).toMatchSnapshot();
});

it('should handle advanced html trees', () => {
  const html =
    '<div id="foo"><ul class="list"><li class="list-item"><a href="/about">About</a></li><li class="list-item"><a href="/contact">Contact</a></li></ul><input type="text" required></div>';
  const result = parseHtml(html);

  expect(result).toEqual([
    {
      type: 'tag',
      name: 'div',
      props: { id: 'foo' },
      children: [
        {
          type: 'tag',
          name: 'ul',
          props: { class: 'list' },
          children: ['About', 'Contact'].map(title => ({
            type: 'tag',
            name: 'li',
            props: { class: 'list-item' },
            children: [
              {
                type: 'tag',
                name: 'a',
                props: { href: `/${title.toLowerCase()}` },
                children: [{ type: 'text', data: title }],
              },
            ],
          })),
        },
        {
          type: 'tag',
          name: 'input',
          props: { type: 'text', required: true },
          children: [],
        },
      ],
    },
  ]);
});
