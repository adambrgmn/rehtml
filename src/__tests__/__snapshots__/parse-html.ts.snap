// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should parse an html string into a DOMTree 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "data": "Hello world",
        "type": "text",
      },
    ],
    "name": "p",
    "props": Object {
      "class": "foo",
    },
    "type": "tag",
  },
]
`;
